File: PAT_Basic_Level_by_lua/1014 福尔摩斯的约会/1014 福尔摩斯的约会.lua
-- 题目编号: 1014 福尔摩斯的约会
-- 实现原理: 依次扫描前两对字符串找到第一对相同的大写字母（A~G）确定星期，再在后续位置找到相同的数字或字母（A~N）确定小时；扫描后两对字符串找到第一对相同的英文字母位置确定分钟

-- 读取4行输入字符串
local a, b, c, d = io.read("l"), io.read("l"), io.read("l"), io.read("l")

-- 星期对照表：A-G对应星期一到星期日
local days = { A = "MON", B = "TUE", C = "WED", D = "THU", E = "FRI", F = "SAT", G = "SUN" }
-- 存储解析出的星期和小时
local day, hour

-- 遍历前两个字符串的对应位置（取较短的长度）
for i = 1, math.min(#a, #b) do
    -- 取出两个字符串当前位置的字符
    local x, y = a:sub(i, i), b:sub(i, i)
    
    -- 还没找到星期，且两个字符相同且是A~G中的一个
    if not day and x == y and days[x] then
        day = days[x]
    -- 已经找到星期，继续找小时：两个字符相同，且是数字0-9或字母A-N
    elseif day and x == y and ((x >= "0" and x <= "9") or (x >= "A" and x <= "N")) then
        -- 数字0-9直接转换，字母A-N对应10-23
        hour = x >= "0" and x <= "9" and tonumber(x) or string.byte(x) - string.byte("A") + 10
        -- 找到小时后跳出循环
        break
    end
end

-- 存储解析出的分钟
local minute
-- 遍历后两个字符串的对应位置
for i = 1, math.min(#c, #d) do
    -- 取出第一个字符串当前位置的字符
    local x = c:sub(i, i)
    -- 如果两个字符相同，且是英文字母，则该位置的下标（从0开始）就是分钟
    if x == d:sub(i, i) and x:match("%a") then
        minute = i - 1
        break
    end
end

-- 按格式输出结果：星期 小时:分钟（小时和分钟都占2位，不足补0）
print(string.format("%s %02d:%02d", day, hour, minute))

File: PAT_Basic_Level_by_ruby/1014 福尔摩斯的约会/1014 福尔摩斯的约会.rb
# 1014 福尔摩斯的约会
#
# 题目描述：
#   大侦探福尔摩斯接到一张奇怪的字条：
#   ```
#   我们约会吧！
#   3485djDkxh4hhGE
#   2984akDfkkkkggEdsb
#   s&hgsfdk
#   d&Hyscvnm
#   ```
#   大侦探很快就明白了，字条上奇怪的乱码实际上就是约会的时间`星期四 14:04`，因为前面两字符串中第 1 对相同的大写英文字母（大小写有区分）是第 4 个字母 `D`，代表星期四；第 2 对相同的字符是 `E` ，那是第 5 个英文字母，代表一天里的第 14 个钟头（于是一天的 0 点到 23 点由数字 0 到 9、以及大写字母 `A` 到 `N` 表示）；后面两字符串第 1 对相同的英文字母 `s` 出现在第 4 个位置（从 0 开始计数）上，代表第 4 分钟。现给定两对字符串，请帮助福尔摩斯解码得到约会的时间。
#
# 实现原理：
#   按特定规则在两个字符串中查找日期信息；依次确定星期、小时、分钟；处理字符的大小写和范围限制
#

DAYS = %w[MON TUE WED THU FRI SAT SUN].freeze

if __FILE__ == $PROGRAM_NAME
  s1 = gets(chomp: true)
  s2 = gets(chomp: true)
  s3 = gets(chomp: true)
  s4 = gets(chomp: true)

  day = nil
  hour = nil
  limit = [s1.length, s2.length].min
  (0...limit).each do |i|
    c1 = s1[i]
    c2 = s2[i]
    next unless c1 == c2

    if day.nil?
      # 第 1 对相同的大写字母 A-G → 星期
      if c1 >= 'A' && c1 <= 'G'
        day = DAYS[c1.ord - 'A'.ord]
        next
      end
    elsif hour.nil?
      # 第 2 对相同字符，且为 0-9 或 A-N → 小时
      if c1 >= '0' && c1 <= '9'
        hour = c1.to_i
        break
      elsif c1 >= 'A' && c1 <= 'N'
        hour = c1.ord - 'A'.ord + 10
        break
      end
    end
  end

  minute = nil
  limit34 = [s3.length, s4.length].min
  (0...limit34).each do |i|
    if s3[i] == s4[i] && s3[i] =~ /[A-Za-z]/
      minute = i
      break
    end
  end

  printf "%s %02d:%02d\n", day, hour, minute
end
